File: src/test/java/com/vjiki/music/service/SongLikeServiceTest.java
package com.vjiki.music.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.vjiki.music.dto.SongLikeResponse;
import com.vjiki.music.entity.AccessLevel;
import com.vjiki.music.entity.AuthProvider;
import com.vjiki.music.entity.Dislike;
import com.vjiki.music.entity.Like;
import com.vjiki.music.entity.Playlist;
import com.vjiki.music.entity.Song;
import com.vjiki.music.entity.User;
import com.vjiki.music.repository.DislikeRepository;
import com.vjiki.music.repository.LikeRepository;
import com.vjiki.music.repository.PlaylistRepository;
import com.vjiki.music.repository.PlaylistSongRepository;
import com.vjiki.music.repository.SongRepository;

@ExtendWith(MockitoExtension.class)
class SongLikeServiceTest {

    private static final String DEFAULT_LIKES = "DEFAULT_LIKES";
    private static final String DEFAULT_DISLIKES = "DEFAULT_DISLIKES";

    @Mock
    private LikeRepository likeRepository;

    @Mock
    private DislikeRepository dislikeRepository;

    @Mock
    private SongRepository songRepository;

    @Mock
    private PlaylistRepository playlistRepository;

    @Mock
    private PlaylistSongRepository playlistSongRepository;

    @InjectMocks
    private SongLikeServiceImpl songLikeService;

    private final UUID userId = UUID.randomUUID();
    private final UUID songId = UUID.randomUUID();

    private User user() {
        return User.builder()
                .id(userId)
                .email("test@example.com")
                .nickname("testuser")
                .provider(AuthProvider.LOCAL)
                .accessLevel(AccessLevel.USER)
                .isActive(true)
                .createdBy("system")
                .modifiedBy("system")
                .build();
    }

    private Song song() {
        return Song.builder()
                .id(songId)
                .artists(Map.of("default", List.of("Artist")))
                .audioUrls(Map.of("default", "http://audio.com"))
                .coverUrls(Map.of("default", "http://cover.com"))
                .title("Test Song")
                .active(true)
                .createdBy("system")
                .modifiedBy("system")
                .build();
    }

    private Playlist playlist(UUID playlistId, String name) {
        return Playlist.builder()
                .id(playlistId)
                .user(user())
                .name(name)
                .type("DEFAULT")
                .isPublic(false)
                .build();
    }

    @Nested
    @DisplayName("likeSong")
    class LikeSong {

        @Test
        @DisplayName("should insert a like and add the song to the default likes playlist")
        void shouldLikeSong() {
            Song song = song();
            when(dislikeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.empty());
            when(songRepository.findById(songId)).thenReturn(Optional.of(song));
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(1L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(0L);

            songLikeService.likeSong(userId, songId);

            verify(likeRepository).insertLike(userId, songId, "system");
            verify(playlistSongRepository).addSongToPlaylistIfNotExists(userId, songId, DEFAULT_LIKES);
            verify(songRepository).save(song);
            assertThat(song.getLikesCount()).isEqualTo(1L);
            assertThat(song.getDislikesCount()).isZero();
        }

        @Test
        @DisplayName("should not touch dislikes when the song was never disliked")
        void shouldNotTouchDislikesWhenNoneExists() {
            when(dislikeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.empty());
            when(songRepository.findById(songId)).thenReturn(Optional.of(song()));
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(1L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(0L);

            songLikeService.likeSong(userId, songId);

            verify(dislikeRepository, never()).save(any());
            verify(playlistRepository, never()).findByUserIdAndName(any(), any());
            verify(playlistSongRepository, never()).deleteByPlaylistIdAndSongId(any(), any());
        }

        @Test
        @DisplayName("should revoke an existing dislike and drop the song from the dislikes playlist")
        void shouldRevokeDislikeWhenLiking() {
            UUID dislikesPlaylistId = UUID.randomUUID();
            Dislike dislike = Dislike.builder()
                    .id(UUID.randomUUID())
                    .user(user())
                    .song(song())
                    .createdBy("system")
                    .build();
            Song song = song();

            when(dislikeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.of(dislike));
            when(playlistRepository.findByUserIdAndName(userId, DEFAULT_DISLIKES))
                    .thenReturn(Optional.of(playlist(dislikesPlaylistId, DEFAULT_DISLIKES)));
            when(songRepository.findById(songId)).thenReturn(Optional.of(song));
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(1L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(0L);

            songLikeService.likeSong(userId, songId);

            assertThat(dislike.getRevokedAt()).isNotNull();
            verify(dislikeRepository).save(dislike);
            verify(playlistSongRepository).deleteByPlaylistIdAndSongId(dislikesPlaylistId, songId);
            verify(likeRepository).insertLike(userId, songId, "system");
        }

        @Test
        @DisplayName("should still insert the like when the song row cannot be found for count refresh")
        void shouldSkipCountUpdateWhenSongMissing() {
            when(dislikeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.empty());
            when(songRepository.findById(songId)).thenReturn(Optional.empty());

            songLikeService.likeSong(userId, songId);

            verify(likeRepository).insertLike(userId, songId, "system");
            verify(songRepository, never()).save(any());
        }
    }

    @Nested
    @DisplayName("dislikeSong")
    class DislikeSong {

        @Test
        @DisplayName("should insert a dislike and add the song to the default dislikes playlist")
        void shouldDislikeSong() {
            Song song = song();
            when(likeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.empty());
            when(songRepository.findById(songId)).thenReturn(Optional.of(song));
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(0L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(1L);

            songLikeService.dislikeSong(userId, songId);

            verify(dislikeRepository).insertDislike(userId, songId, "system");
            verify(playlistSongRepository).addSongToPlaylistIfNotExists(userId, songId, DEFAULT_DISLIKES);
            verify(songRepository).save(song);
            assertThat(song.getLikesCount()).isZero();
            assertThat(song.getDislikesCount()).isEqualTo(1L);
        }

        @Test
        @DisplayName("should not touch likes when the song was never liked")
        void shouldNotTouchLikesWhenNoneExists() {
            when(likeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.empty());
            when(songRepository.findById(songId)).thenReturn(Optional.of(song()));
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(0L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(1L);

            songLikeService.dislikeSong(userId, songId);

            verify(likeRepository, never()).save(any());
            verify(playlistRepository, never()).findByUserIdAndName(any(), any());
            verify(playlistSongRepository, never()).deleteByPlaylistIdAndSongId(any(), any());
        }

        @Test
        @DisplayName("should revoke an existing like and drop the song from the likes playlist")
        void shouldRevokeLikeWhenDisliking() {
            UUID likesPlaylistId = UUID.randomUUID();
            Like like = Like.builder()
                    .id(UUID.randomUUID())
                    .user(user())
                    .song(song())
                    .createdBy("system")
                    .build();

            when(likeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.of(like));
            when(playlistRepository.findByUserIdAndName(userId, DEFAULT_LIKES))
                    .thenReturn(Optional.of(playlist(likesPlaylistId, DEFAULT_LIKES)));
            when(songRepository.findById(songId)).thenReturn(Optional.of(song()));
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(0L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(1L);

            songLikeService.dislikeSong(userId, songId);

            assertThat(like.getRevokedAt()).isNotNull();
            verify(likeRepository).save(like);
            verify(playlistSongRepository).deleteByPlaylistIdAndSongId(likesPlaylistId, songId);
            verify(dislikeRepository).insertDislike(userId, songId, "system");
        }

        @Test
        @DisplayName("should still insert the dislike when the song row cannot be found for count refresh")
        void shouldSkipCountUpdateWhenSongMissing() {
            when(likeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId))
                    .thenReturn(Optional.empty());
            when(songRepository.findById(songId)).thenReturn(Optional.empty());

            songLikeService.dislikeSong(userId, songId);

            verify(dislikeRepository).insertDislike(userId, songId, "system");
            verify(songRepository, never()).save(any());
        }
    }

    @Nested
    @DisplayName("getLikeDislikeInfo")
    class GetLikeDislikeInfo {

        @Test
        @DisplayName("should return correct like/dislike info")
        void shouldReturnLikeDislikeInfo() {
            when(likeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId)).thenReturn(true);
            when(dislikeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId)).thenReturn(false);
            when(likeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(10L);
            when(dislikeRepository.countBySongIdAndRevokedAtIsNull(songId)).thenReturn(2L);

            SongLikeResponse result = songLikeService.getLikeDislikeInfo(userId, songId);

            assertThat(result.getIsLiked()).isTrue();
            assertThat(result.getIsDisliked()).isFalse();
            assertThat(result.getLikesCount()).isEqualTo(10L);
            assertThat(result.getDislikesCount()).isEqualTo(2L);
        }
    }
}

File: src/test/kotlin/com/vjiki/music/service/SongLikeServiceTest.kt
package com.vjiki.music.service

import com.vjiki.music.dto.SongLikeResponse
import com.vjiki.music.entity.*
import com.vjiki.music.repository.*
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.*
import org.springframework.data.repository.findByIdOrNull
import java.time.OffsetDateTime
import java.util.*

class SongLikeServiceTest : DescribeSpec({

    val likeRepository = mockk<LikeRepository>()
    val dislikeRepository = mockk<DislikeRepository>()
    val songRepository = mockk<SongRepository>()
    val userRepository = mockk<UserRepository>()
    val playlistRepository = mockk<PlaylistRepository>()
    val playlistSongRepository = mockk<PlaylistSongRepository>()

    val songLikeService = SongLikeServiceImpl(
        likeRepository,
        dislikeRepository,
        songRepository,
        userRepository,
        playlistRepository,
        playlistSongRepository
    )

    describe("likeSong") {
        val userId = UUID.randomUUID()
        val songId = UUID.randomUUID()
        val user = User(
            id = userId,
            email = "test@example.com",
            nickname = "testuser",
            provider = AuthProvider.LOCAL,
            accessLevel = AccessLevel.USER,
            isActive = true,
            createdBy = "system",
            modifiedBy = "system"
        )
        val song = Song(
            id = songId,
            artists = mapOf("default" to listOf("Artist")),
            audioUrls = mapOf("default" to "http://audio.com"),
            coverUrls = mapOf("default" to "http://cover.com"),
            title = "Test Song",
            active = true,
            createdBy = "system",
            modifiedBy = "system"
        )

        it("should like a song when not already liked") {
            val playlistId = UUID.randomUUID()
            val playlist = Playlist(
                id = playlistId,
                user = user,
                name = "DEFAULT_LIKES",
                type = "DEFAULT",
                isPublic = false,
                createdBy = "system",
                modifiedBy = "system"
            )

            every { likeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns false
            every { dislikeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns Optional.empty()
            every { userRepository.findById(userId) } returns Optional.of(user)
            every { songRepository.findById(songId) } returns Optional.of(song)
            every { likeRepository.save(any()) } returns mockk()
            every { playlistRepository.findByUserIdAndName(userId, "DEFAULT_LIKES") } returns Optional.empty()
            every { playlistRepository.save(any()) } returns playlist
            every { playlistSongRepository.findByPlaylistIdAndSongId(playlistId, songId) } returns Optional.empty()
            every { playlistSongRepository.findByPlaylistIdWithSong(playlistId) } returns emptyList()
            every { playlistSongRepository.save(any()) } returns mockk()
            every { likeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 1L
            every { dislikeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 0L
            every { songRepository.save(any()) } returns song

            songLikeService.likeSong(userId, songId)

            verify(exactly = 1) { likeRepository.save(any()) }
            verify(exactly = 1) { songRepository.save(any()) }
        }

        it("should not like if already liked") {
            clearMocks(likeRepository, dislikeRepository, userRepository, songRepository, playlistRepository, playlistSongRepository)
            every { likeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns true

            songLikeService.likeSong(userId, songId)

            verify(exactly = 0) { likeRepository.save(any()) }
            verify(exactly = 0) { userRepository.findById(any()) }
            verify(exactly = 0) { songRepository.findById(any()) }
        }

        it("should revoke dislike when liking") {
            val dislike = Dislike(
                user = user,
                song = song,
                createdBy = "system"
            )
            val dislikesPlaylistId = UUID.randomUUID()
            val dislikesPlaylist = Playlist(
                id = dislikesPlaylistId,
                user = user,
                name = "DEFAULT_DISLIKES",
                type = "DEFAULT",
                isPublic = false,
                createdBy = "system",
                modifiedBy = "system"
            )
            val likesPlaylistId = UUID.randomUUID()
            val likesPlaylist = Playlist(
                id = likesPlaylistId,
                user = user,
                name = "DEFAULT_LIKES",
                type = "DEFAULT",
                isPublic = false,
                createdBy = "system",
                modifiedBy = "system"
            )

            every { likeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns false
            every { dislikeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns Optional.of(dislike)
            every { dislikeRepository.save(any()) } returns dislike
            every { playlistRepository.findByUserIdAndName(userId, "DEFAULT_DISLIKES") } returns Optional.of(dislikesPlaylist)
            every { playlistSongRepository.deleteByPlaylistIdAndSongId(dislikesPlaylistId, songId) } just Runs
            every { userRepository.findById(userId) } returns Optional.of(user)
            every { songRepository.findById(songId) } returns Optional.of(song)
            every { likeRepository.save(any()) } returns mockk()
            every { playlistRepository.findByUserIdAndName(userId, "DEFAULT_LIKES") } returns Optional.empty()
            every { playlistRepository.save(any()) } returns likesPlaylist
            every { playlistSongRepository.findByPlaylistIdAndSongId(likesPlaylistId, songId) } returns Optional.empty()
            every { playlistSongRepository.findByPlaylistIdWithSong(likesPlaylistId) } returns emptyList()
            every { playlistSongRepository.save(any()) } returns mockk()
            every { likeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 1L
            every { dislikeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 0L
            every { songRepository.save(any()) } returns song

            songLikeService.likeSong(userId, songId)

            verify(exactly = 1) { dislikeRepository.save(any()) }
            verify(exactly = 1) { likeRepository.save(any()) }
        }
    }

    describe("dislikeSong") {
        val userId = UUID.randomUUID()
        val songId = UUID.randomUUID()
        val user = User(
            id = userId,
            email = "test@example.com",
            nickname = "testuser",
            provider = AuthProvider.LOCAL,
            accessLevel = AccessLevel.USER,
            isActive = true,
            createdBy = "system",
            modifiedBy = "system"
        )
        val song = Song(
            id = songId,
            artists = mapOf("default" to listOf("Artist")),
            audioUrls = mapOf("default" to "http://audio.com"),
            coverUrls = mapOf("default" to "http://cover.com"),
            title = "Test Song",
            active = true,
            createdBy = "system",
            modifiedBy = "system"
        )

        it("should dislike a song when not already disliked") {
            clearMocks(likeRepository, dislikeRepository, userRepository, songRepository, playlistRepository, playlistSongRepository)
            val playlistId = UUID.randomUUID()
            val playlist = Playlist(
                id = playlistId,
                user = user,
                name = "DEFAULT_DISLIKES",
                type = "DEFAULT",
                isPublic = false,
                createdBy = "system",
                modifiedBy = "system"
            )

            every { dislikeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns false
            every { likeRepository.findByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns Optional.empty()
            every { userRepository.findById(userId) } returns Optional.of(user)
            every { songRepository.findById(songId) } returns Optional.of(song)
            every { dislikeRepository.save(any()) } returns mockk()
            every { playlistRepository.findByUserIdAndName(userId, "DEFAULT_DISLIKES") } returns Optional.empty()
            every { playlistRepository.save(any()) } returns playlist
            every { playlistSongRepository.findByPlaylistIdAndSongId(playlistId, songId) } returns Optional.empty()
            every { playlistSongRepository.findByPlaylistIdWithSong(playlistId) } returns emptyList()
            every { playlistSongRepository.save(any()) } returns mockk()
            every { likeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 0L
            every { dislikeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 1L
            every { songRepository.save(any()) } returns song

            songLikeService.dislikeSong(userId, songId)

            verify(exactly = 1) { dislikeRepository.save(any()) }
            verify(exactly = 1) { songRepository.save(any()) }
        }

        it("should not dislike if already disliked") {
            clearMocks(likeRepository, dislikeRepository, userRepository, songRepository, playlistRepository, playlistSongRepository)
            every { dislikeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns true

            songLikeService.dislikeSong(userId, songId)

            verify(exactly = 0) { dislikeRepository.save(any()) }
            verify(exactly = 0) { userRepository.findById(any()) }
            verify(exactly = 0) { songRepository.findById(any()) }
        }
    }

    describe("getLikeDislikeInfo") {
        val userId = UUID.randomUUID()
        val songId = UUID.randomUUID()

        it("should return correct like/dislike info") {
            every { likeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns true
            every { dislikeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(userId, songId) } returns false
            every { likeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 10L
            every { dislikeRepository.countBySongIdAndRevokedAtIsNull(songId) } returns 2L

            val result = songLikeService.getLikeDislikeInfo(userId, songId)

            result.isLiked shouldBe true
            result.isDisliked shouldBe false
            result.likesCount shouldBe 10L
            result.dislikesCount shouldBe 2L
        }
    }
})

